File: tools/usage-poc/coconut_collector/collect.py
"""collect.py — build a Burn Summary envelope from local session logs.

Groups parsed sessions by (tool, model, projectHash) and emits an envelope
conforming to web/tools/usage-poc/burn-summary.schema.json.

SECURITY: output carries ONLY the 9 uploadable fields (handoff §8). Project
path slugs are consumed by project_hash() and never emitted; content fields
are never read.
"""

import json
import re
from datetime import datetime, timedelta, timezone
from pathlib import Path

from .gitcount import count_commits, git_author_email
from .hashing import load_or_create_salt
from .parsers import (cost_breakdown, find_logs, load_pricing, match_model,
                      parse_claude, parse_codex)

# PoC-internal tool name -> Burn Summary `tool` enum.
_TOOL_NAME = {"claude": "claude-code", "codex": "codex"}
_DAY_RE = re.compile(r"^(\d{4}-\d{2}-\d{2})")

# Selectable leaderboard windows. 'all' disables filtering (local audit only).
_PERIODS = ("day", "week", "month", "year", "all")

# Sentinel added to the cwd sink for an in-window, token-contributing session
# that carries NO cwd. Such work is unattributable to a repo, so count_commits
# SKIPS it (the sentinel resolves to no git root) — the numerator counts only
# the verifiable repos (conservative undercount, see gitcount.py PARTIAL
# ATTRIBUTION). The NUL byte cannot occur in a real filesystem path, so it
# never collides with a genuine cwd.
_UNKNOWN_CWD = "\x00no-cwd"


def _utc_day(timestamp: str | None) -> str | None:
    """Extract a 'YYYY-MM-DD' UTC day from an ISO-8601 timestamp."""
    if not timestamp:
        return None
    m = _DAY_RE.match(timestamp)
    return m.group(1) if m else None


def _parse_instant(timestamp: str | None) -> datetime | None:
    """Parse an ISO-8601 session timestamp into an aware UTC datetime.

    'Z' is normalised to '+00:00'; naive timestamps are assumed UTC.
    Returns None when the timestamp is missing or unparseable.
    """
    if not timestamp:
        return None
    try:
        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
    except ValueError:
        return None
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def _calendar_window(period: str,
                     now: datetime) -> tuple[datetime, datetime] | None:
    """Return the [since, until) UTC calendar bucket for `now`.

    'all' returns None (no filtering). Buckets are calendar-aligned, not
    rolling: day/month/year snap to the bucket containing `now`. 'week' is
    the LAST COMPLETED ISO week (prior Monday..this Monday), not the
    in-progress one — so a Monday and a Sunday importer compete over the
    same fully-elapsed 7 days.
    """
    if period == "all":
        return None
    day0 = now.astimezone(timezone.utc).replace(
        hour=0, minute=0, second=0, microsecond=0)
    if period == "day":
        return day0, day0 + timedelta(days=1)
    if period == "week":
        this_monday = day0 - timedelta(days=day0.weekday())
        return this_monday - timedelta(weeks=1), this_monday
    if period == "month":
        since = day0.replace(day=1)
        if since.month == 12:
            until = since.replace(year=since.year + 1, month=1)
        else:
            until = since.replace(month=since.month + 1)
        return since, until
    if period == "year":
        since = day0.replace(month=1, day=1)
        return since, since.replace(year=since.year + 1)
    raise ValueError(f"unknown period: {period!r}")


def _in_window(sp, window: tuple[datetime, datetime] | None) -> bool:
    """Whether session `sp` belongs to `window` (attributed by start time).

    None window means no filtering. A session whose timestamp cannot be
    parsed is excluded when a window is active — it cannot prove membership.
    """
    if window is None:
        return True
    instant = _parse_instant(sp.timestamp)
    if instant is None:
        return False
    since, until = window
    return since <= instant < until


def _schema_token_count(tool: str, tok: dict) -> dict:
    """Map a PoC tok dict to the schema's tokenCount shape."""
    if tool == "claude":
        return {
            "input": tok.get("input", 0),
            "output": tok.get("output", 0),
            "cacheRead": tok.get("cache_read", 0),
            "cacheWrite": tok.get("cache_write_5m", 0) + tok.get("cache_write_1h", 0),
            "cachedInput": 0,
        }
    return {
        "input": tok.get("input", 0),
        "output": tok.get("output", 0),
        "cacheRead": 0,
        "cacheWrite": 0,
        "cachedInput": tok.get("cached_input", 0),
    }


def _verification(price_confidence: str) -> dict:
    """Build the verification object for a device-collected, estimated row.

    Unmatched model (low confidence) downgrades the display level to
    'Estimated' — see coconutlabs-verification-model.md §0-A.
    """
    level = "Device-synced" if price_confidence == "high" else "Estimated"
    return {
        "tokenSource": "device",
        "costBasis": "estimated",
        "priceConfidence": price_confidence,
        "level": level,
    }


def collect(pricing: dict, salt: str, period: str = "all",
            now: datetime | None = None,
            scan_root: "Path | None" = None,
            cwd_sink: "set | None" = None) -> dict:
    """Scan every local session log and aggregate into grouped rows.

    Each log file is one session. Files that fail to parse or carry zero
    tokens are skipped. When `period` is not 'all', sessions are filtered
    to the calendar window containing `now` (defaults to the current UTC
    instant); a session is attributed by its first-line start timestamp.

    `scan_root` overrides the default log discovery paths. When supplied,
    logs are searched under ``scan_root/.claude/projects`` and
    ``scan_root/.codex/sessions`` instead of the standard home-relative
    defaults. Useful when the user passes their home directory explicitly
    (e.g. ``coconut-collector ~/``).

    `cwd_sink`, when provided, is populated with the raw `cwd` of every
    in-window session that has one; a token-contributing session that lacks a
    cwd adds the `_UNKNOWN_CWD` sentinel instead. count_commits SKIPS the
    sentinel (and any non-git cwd) rather than poisoning — it counts only the
    verifiable repos (conservative undercount). The caller uses the sink ONLY
    locally to resolve git repo roots for the commit count — cwds never enter
    the emitted envelope.
    """
    if now is None:
        now = datetime.now(timezone.utc)
    window = _calendar_window(period, now)
    groups: dict[tuple, dict] = {}
    for tool in ("claude", "codex"):
        parse = parse_claude if tool == "claude" else parse_codex
        for path in find_logs(tool, scan_root=scan_root):
            try:
                sp = parse(path, salt)
            except (ValueError, OSError, json.JSONDecodeError):
                continue
            if sum(sp.tokens.values()) == 0:
                continue
            if not _in_window(sp, window):
                continue
            if cwd_sink is not None:
                # A contributing session with no cwd is work we can't attribute
                # to a repo → record the _UNKNOWN_CWD sentinel, which
                # count_commits SKIPS (conservative undercount, not a poison).
                cwd_sink.add(sp.cwd if sp.cwd else _UNKNOWN_CWD)
            key = (tool, sp.model, sp.project_hash)
            grp = groups.get(key)
            if grp is None:
                grp = {"tokens": {k: 0 for k in sp.tokens},
                       "sessions": 0, "days": set()}
                groups[key] = grp
            for k, v in sp.tokens.items():
                grp["tokens"][k] = grp["tokens"].get(k, 0) + v
            grp["sessions"] += 1
            day = _utc_day(sp.timestamp)
            if day:
                grp["days"].add(day)
    return groups


def build_envelope(pricing: dict, salt: str,
                   generated_at: str | None = None,
                   period: str = "week",
                   scan_root: "Path | None" = None) -> dict:
    """Assemble the Burn Summary envelope (schemaVersion 3).

    `period` selects the calendar window (day/week/month/year/all). The
    window end and `generatedAt` are anchored to the same instant. Raises
    ValueError when `period` is unknown or no sessions fall in the window.
    `scan_root` is forwarded to `collect()` to override default log paths.
    """
    if period not in _PERIODS:
        raise ValueError(f"unknown period: {period!r}")
    if generated_at is None:
        now = datetime.now(timezone.utc)
        generated_at = now.strftime("%Y-%m-%dT%H:%M:%SZ")
    else:
        now = _parse_instant(generated_at)
        if now is None:
            raise ValueError(f"invalid generatedAt: {generated_at!r}")
        # Re-derive from the parsed instant so an offset/fractional-second
        # input still serialises as schema-valid second-precision UTC 'Z'.
        generated_at = now.strftime("%Y-%m-%dT%H:%M:%SZ")
    fallback_day = generated_at[:10]
    repo_cwds: set[str] = set()
    groups = collect(pricing, salt, period=period, now=now,
                     scan_root=scan_root, cwd_sink=repo_cwds)
    rows = []
    total_tokens = 0
    total_cost = 0.0
    for (tool, model, phash), grp in sorted(groups.items()):
        tok = grp["tokens"]
        price, confidence = match_model(pricing.get(tool, {}), model)
        cost = round(sum(cost_breakdown(tok, price).values()), 4)
        token_count = _schema_token_count(tool, tok)
        row_total = sum(token_count.values())
        days = sorted(grp["days"])
        rows.append({
            "tool": _TOOL_NAME[tool],
            "model": model,
            "tokenCount": token_count,
            "estimatedCostUsd": cost,
            "timestampBucket": days[-1] if days else fallback_day,
            "sessionCount": grp["sessions"],
            "activeDays": len(days),
            "projectHash": phash,
            "verification": _verification(confidence),
        })
        total_tokens += row_total
        total_cost += cost
    if not rows:
        raise ValueError(f"no sessions in period '{period}'")
    window = _calendar_window(period, now)
    if window is None:
        since_iso = until_iso = None
    else:
        since_iso = window[0].strftime("%Y-%m-%dT%H:%M:%SZ")
        until_iso = window[1].strftime("%Y-%m-%dT%H:%M:%SZ")
    envelope = {
        "schemaVersion": "3",
        "generatedAt": generated_at,
        "periodWindow": {
            "period": period,
            "since": since_iso,
            "until": until_iso,
        },
        "rows": rows,
        "grandTotal": {
            "totalTokens": total_tokens,
            "estimatedCostUsd": round(total_cost, 4),
        },
    }
    # VES numerator: device-measured commit count over the SAME window.
    # Emitted only when bounded ('all' has no window) and the count is known
    # (count_commits returns None when NO cwd resolves to a git repo, or a
    # resolved repo errors out — see gitcount.py); a real 0 is emitted. Non-git
    # and cwd-less sessions are skipped inside count_commits (conservative
    # undercount), so the sentinel no longer gates emission here. cwd/SHA/email
    # never leave this process.
    if window is not None and repo_cwds:
        vc = count_commits(repo_cwds, git_author_email(), window[0], window[1])
        if vc is not None:
            envelope["verifiedCommits"] = vc
    return envelope


def print_table(envelope: dict) -> None:
    """Human-readable summary (aggregates only)."""
    print("=== CoconutLabs Burn Summary ===")
    print(f"  schemaVersion:  {envelope['schemaVersion']}")
    print(f"  generatedAt:    {envelope['generatedAt']}")
    pw = envelope.get("periodWindow")
    if pw:
        span = (f"{pw['since']} .. {pw['until']}"
                if pw["since"] else "(unfiltered)")
        print(f"  period:         {pw['period']}  [{span}]")
    print(f"  rows:           {len(envelope['rows'])}")
    print("  --- per group (tool / model / projectHash) ---")
    for r in envelope["rows"]:
        v = r["verification"]
        row_total = sum(r["tokenCount"].values())
        print(f"  [{r['tool']}] {r['model']}  {r['projectHash']}  ({v['level']})")
        print(f"    {row_total:>15,} tokens"
              f"    ${r['estimatedCostUsd']:>12.4f}"
              f"    {r['sessionCount']} sessions / {r['activeDays']} active days")
    gt = envelope["grandTotal"]
    print("  --- grand total ---")
    print(f"    {gt['totalTokens']:>15,} tokens"
          f"    ${gt['estimatedCostUsd']:>12.4f}")

File: lib/client/burn/collect.ts
// collect.ts — assemble a Burn Summary envelope from FSA-discovered logs.
//
// Browser port of coconut_collector/collect.py. Groups parsed sessions by
// (tool, model, projectHash) and emits an envelope conforming to
// web/tools/usage-poc/burn-summary.schema.json.
//
// SECURITY (mirrors collect.py header): output carries ONLY the 9
// uploadable fields. Project slugs are consumed by projectHash() and never
// emitted. Content fields are never read by parsers.ts; this module never
// looks at raw line text at all — it operates on already-derived
// SessionParse scalars.
//
// Parity risks vs the Python reference, each addressed in this file:
//   1) Python `round(x, 4)` is banker's rounding (half-to-even); JS
//      Math.round is half-away-from-zero. We use bankersRound() so a row
//      whose true cost is 0.12345 lands on 0.1234 (matching Python) instead
//      of 0.1235.
//   2) `_utc_day` extracts a 'YYYY-MM-DD' prefix from the raw timestamp
//      string — it is NOT timezone-normalised. We mirror that with a regex
//      slice rather than constructing a Date.
//   3) `_parse_instant` treats naive timestamps as UTC. `new Date(s)` in JS
//      varies (Safari: local time, Chrome: UTC) for offset-less strings, so
//      we explicitly append 'Z' before parsing when no offset is present.
//   4) ISO weekday: Python's `weekday()` is Mon=0..Sun=6, JS getUTCDay() is
//      Sun=0..Sat=6. We compute Mon-aligned weekday with
//      `(getUTCDay() + 6) % 7` for the LAST COMPLETED ISO week math.

import {
  costBreakdown,
  findClaudeLogs,
  findCodexLogs,
  matchModel,
  parseClaudeFile,
  parseCodexFile,
  safeAdd,
  type ClaudeTokens,
  type CodexTokens,
  type SessionParse,
  type SessionTokens,
} from "./parsers";
import { projectHash } from "./hashing";
import {
  MODEL_PRICING,
  type PricingTable,
  type ClaudeRate,
  type CodexRate,
} from "./pricing.generated";

// PoC-internal tool name → Burn Summary `tool` enum (mirrors collect.py).
const TOOL_NAME = { claude: "claude-code", codex: "codex" } as const;

// Selectable leaderboard windows. 'all' disables filtering (local audit
// only — Burn Summary submission always uses 'week').
export type Period = "day" | "week" | "month" | "year" | "all";
const PERIODS: readonly Period[] = ["day", "week", "month", "year", "all"];

// Mirrors collect.py _DAY_RE.
const DAY_RE = /^(\d{4}-\d{2}-\d{2})/;

// Detects a trailing offset on an ISO-8601 string ('Z', '+HH:MM', '-HH:MM',
// '+HHMM', '-HHMM'). Used to decide whether to append 'Z' so a naive
// timestamp is parsed as UTC across all browsers.
const TZ_OFFSET_RE = /(Z|[+-]\d{2}:?\d{2})$/;

// CPython parity for `round(x, 4)`. The cost field round-trips through the
// schema and drives leaderboard ranking — a single-cent drift between the
// Python collector and the TS browser collector would let two identical
// sessions hash to different rows, splitting projects across the table.
//
// CPython's `round(x, n)` calls `_Py_dg_dtoa(x, mode=3, ndigits=n)` which
// rounds the ACTUAL IEEE-754 bit pattern of x to n decimals. Because
// 0.00025 is stored as 0.00025000000000000000520... (slightly LARGER than
// the exact decimal), CPython rounds it UP to 0.0003 — even though pure
// banker's rounding on the decimal string "0.00025" would round to 0.0002.
//
// A prior version of this function used `toPrecision(15)` to "normalise"
// representation noise before banker's rounding. That destroyed the very
// bit pattern CPython relies on, producing systematic divergence (0.00025
// → 0.0002 instead of 0.0003, 0.12345 → 0.1234 instead of 0.1235, etc.).
//
// V8's `Number.prototype.toFixed(decimals)` happens to match CPython's
// behaviour exactly: it uses double-conversion against the raw bit pattern
// just like `_Py_dg_dtoa`. Empirically verified parity on 3015 test cases
// (cost-like decimals × 10⁻⁶..10⁻³ range + half-way edges + negatives +
// signed zero). `Number()` strips trailing zeros so `2.6750` → `2.675`.
function bankersRound(value: number, decimals: number): number {
  if (!Number.isFinite(value)) return value;
  return Number(value.toFixed(decimals));
}

// Mirrors collect.py _utc_day: regex-slice the date prefix from the raw
// ISO string. NOT timezone-normalised — a '2026-05-20T23:30:00-05:00'
// timestamp counts as 2026-05-20, same as Python.
function utcDay(timestamp: string | null): string | null {
  if (!timestamp) return null;
  const m = DAY_RE.exec(timestamp);
  return m ? m[1] : null;
}

// Mirrors collect.py _parse_instant. Returns a Date in UTC, or null when
// the string is missing or unparseable. Naive timestamps (no offset) are
// treated as UTC by appending 'Z' before construction — `new Date(s)` is
// implementation-defined for offset-less strings.
function parseInstant(timestamp: string | null): Date | null {
  if (!timestamp) return null;
  const hasOffset = TZ_OFFSET_RE.test(timestamp);
  const normalised = hasOffset ? timestamp : `${timestamp}Z`;
  const dt = new Date(normalised);
  return Number.isNaN(dt.getTime()) ? null : dt;
}

// ISO weekday: Monday=0..Sunday=6. JS getUTCDay() is Sun=0..Sat=6, so we
// shift by 6 and mod 7 to land Monday at 0.
function isoWeekday(date: Date): number {
  return (date.getUTCDay() + 6) % 7;
}

// Format a Date as schema-valid second-precision UTC 'Z'. Mirrors Python's
// strftime("%Y-%m-%dT%H:%M:%SZ"). toISOString() emits millisecond precision
// (e.g. '2026-05-20T12:34:56.789Z'), which the schema rejects.
function formatGenerated(date: Date): string {
  const y = date.getUTCFullYear().toString().padStart(4, "0");
  const mo = (date.getUTCMonth() + 1).toString().padStart(2, "0");
  const d = date.getUTCDate().toString().padStart(2, "0");
  const h = date.getUTCHours().toString().padStart(2, "0");
  const mi = date.getUTCMinutes().toString().padStart(2, "0");
  const s = date.getUTCSeconds().toString().padStart(2, "0");
  return `${y}-${mo}-${d}T${h}:${mi}:${s}Z`;
}

// Snap a UTC instant to the start of its day.
function utcDayStart(date: Date): Date {
  return new Date(Date.UTC(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    0, 0, 0, 0,
  ));
}

// Mirrors collect.py _calendar_window. Returns null for 'all' (no
// filtering), or a [since, until) UTC pair. 'week' is the LAST COMPLETED
// ISO week — Monday-aligned, [prior_monday, this_monday) — so a Monday and
// a Sunday importer compete over the same fully-elapsed 7 days.
function calendarWindow(period: Period, now: Date): [Date, Date] | null {
  if (period === "all") return null;
  const day0 = utcDayStart(now);
  const MS_DAY = 86_400_000;
  if (period === "day") {
    return [day0, new Date(day0.getTime() + MS_DAY)];
  }
  if (period === "week") {
    const thisMonday = new Date(day0.getTime() - isoWeekday(day0) * MS_DAY);
    const priorMonday = new Date(thisMonday.getTime() - 7 * MS_DAY);
    return [priorMonday, thisMonday];
  }
  if (period === "month") {
    const since = new Date(Date.UTC(
      day0.getUTCFullYear(),
      day0.getUTCMonth(),
      1,
      0, 0, 0, 0,
    ));
    const until = since.getUTCMonth() === 11
      ? new Date(Date.UTC(since.getUTCFullYear() + 1, 0, 1, 0, 0, 0, 0))
      : new Date(Date.UTC(
          since.getUTCFullYear(),
          since.getUTCMonth() + 1,
          1,
          0, 0, 0, 0,
        ));
    return [since, until];
  }
  if (period === "year") {
    const since = new Date(Date.UTC(day0.getUTCFullYear(), 0, 1, 0, 0, 0, 0));
    const until = new Date(Date.UTC(day0.getUTCFullYear() + 1, 0, 1, 0, 0, 0, 0));
    return [since, until];
  }
  throw new Error(`unknown period: ${period}`);
}

function inWindow(
  sp: SessionParse<SessionTokens>,
  window: [Date, Date] | null,
): boolean {
  if (window === null) return true;
  const instant = parseInstant(sp.timestamp);
  if (instant === null) return false;
  const [since, until] = window;
  const t = instant.getTime();
  return t >= since.getTime() && t < until.getTime();
}

// Map a tool-specific token dict to the schema's tokenCount shape. Mirrors
// collect.py _schema_token_count exactly: claude folds the two cache_write
// tiers into a single field, codex routes cached_input into cachedInput.
interface SchemaTokenCount {
  input: number;
  output: number;
  cacheRead: number;
  cacheWrite: number;
  cachedInput: number;
}

function schemaTokenCount(
  tool: "claude" | "codex",
  tok: SessionTokens,
): SchemaTokenCount {
  if (tool === "claude") {
    const t = tok as ClaudeTokens;
    return {
      input: t.input,
      output: t.output,
      cacheRead: t.cache_read,
      cacheWrite: safeAdd(t.cache_write_5m, t.cache_write_1h),
      cachedInput: 0,
    };
  }
  const t = tok as CodexTokens;
  return {
    input: t.input,
    output: t.output,
    cacheRead: 0,
    cacheWrite: 0,
    cachedInput: t.cached_input,
  };
}

interface Verification {
  tokenSource: "device";
  costBasis: "estimated";
  priceConfidence: "high" | "low";
  level: "Device-synced" | "Estimated";
}

function buildVerification(confidence: "high" | "low"): Verification {
  return {
    tokenSource: "device",
    costBasis: "estimated",
    priceConfidence: confidence,
    level: confidence === "high" ? "Device-synced" : "Estimated",
  };
}

// Aggregated group keyed by (tool, model, projectHash).
interface Group {
  tokens: Record<string, number>;
  sessions: number;
  days: Set<string>;
}

// Sort key: (tool, model, projectHash) lexicographic, matching Python's
// `sorted(groups.items())` which compares tuples element-by-element.
function compareKey(
  a: readonly [string, string, string],
  b: readonly [string, string, string],
): number {
  for (let i = 0; i < 3; i++) {
    if (a[i] < b[i]) return -1;
    if (a[i] > b[i]) return 1;
  }
  return 0;
}

// Convert SessionTokens (a union) to a generic numeric record so the
// aggregator can sum any tier without per-tool branching.
function tokensAsRecord(tokens: SessionTokens): Record<string, number> {
  return tokens as unknown as Record<string, number>;
}

// Handles passed in from the UI layer. Either may be null — a Safari user
// importing only Codex logs, say. At least one must be present.
export interface CollectInputs {
  claudeProjectsHandle: FileSystemDirectoryHandle | null;
  codexSessionsHandle: FileSystemDirectoryHandle | null;
  salt: string;
  now?: Date;
  period?: Period;
}

// Internal: walk handles, parse each session, group by (tool, model, phash).
async function collectGroups(
  inputs: CollectInputs,
  window: [Date, Date] | null,
): Promise<Map<string, Group> & { __keys: Map<string, [string, string, string]> }> {
  const groups = new Map<string, Group>();
  const keyMap = new Map<string, [string, string, string]>();
  const { claudeProjectsHandle, codexSessionsHandle, salt } = inputs;

  // Closure-scoped hash callback. Raw slugs never leave the parser layer —
  // they enter as directory names or `cwd` strings, get hashed here, and
  // only the 12-char hex tail crosses back into collect.ts. The salt is
  // captured by this closure and never written into envelope output.
  const hashSlug = (slug: string): Promise<string> => projectHash(slug, salt);

  async function ingest(
    sp: SessionParse<SessionTokens>,
    tool: "claude" | "codex",
  ): Promise<void> {
    // Skip empty-token sessions (mirrors `sum(sp.tokens.values()) == 0`).
    const totals = Object.values(tokensAsRecord(sp.tokens));
    let sum = 0;
    for (const v of totals) sum += v;
    if (sum === 0) return;
    if (!inWindow(sp, window)) return;
    const phash = sp.projectHash;
    const keyTuple: [string, string, string] = [tool, sp.model, phash];
    const keyStr = `${tool}${sp.model}${phash}`;
    let grp = groups.get(keyStr);
    if (grp === undefined) {
      const seed: Record<string, number> = {};
      for (const k of Object.keys(tokensAsRecord(sp.tokens))) seed[k] = 0;
      grp = { tokens: seed, sessions: 0, days: new Set() };
      groups.set(keyStr, grp);
      keyMap.set(keyStr, keyTuple);
    }
    for (const [k, v] of Object.entries(tokensAsRecord(sp.tokens))) {
      grp.tokens[k] = safeAdd(grp.tokens[k] ?? 0, v);
    }
    grp.sessions += 1;
    const day = utcDay(sp.timestamp);
    if (day) grp.days.add(day);
  }

  if (claudeProjectsHandle) {
    const entries = await findClaudeLogs(claudeProjectsHandle, hashSlug);
    for (const entry of entries) {
      let sp: SessionParse<ClaudeTokens>;
      try {
        sp = await parseClaudeFile(entry.file, entry.projectHash);
      } catch {
        // Per security comments: never log raw paths or line content. A
        // malformed file is silently skipped, same as Python.
        continue;
      }
      await ingest(sp, "claude");
    }
  }

  if (codexSessionsHandle) {
    const entries = await findCodexLogs(codexSessionsHandle);
    for (const entry of entries) {
      let sp: SessionParse<CodexTokens>;
      try {
        sp = await parseCodexFile(entry.file, hashSlug);
      } catch {
        continue;
      }
      await ingest(sp, "codex");
    }
  }

  // Attach key map for downstream sorting without re-parsing the joined key.
  (groups as Map<string, Group> & { __keys: Map<string, [string, string, string]> })
    .__keys = keyMap;
  return groups as Map<string, Group> & {
    __keys: Map<string, [string, string, string]>;
  };
}

// Public envelope row shape — mirrors burn-summary.schema.json.
export interface EnvelopeRow {
  tool: "claude-code" | "codex";
  model: string;
  tokenCount: SchemaTokenCount;
  estimatedCostUsd: number;
  timestampBucket: string;
  sessionCount: number;
  activeDays: number;
  projectHash: string;
  verification: Verification;
}

export interface PeriodWindow {
  period: Period;
  since: string | null;
  until: string | null;
}

export interface Envelope {
  schemaVersion: "2";
  generatedAt: string;
  periodWindow: PeriodWindow;
  rows: EnvelopeRow[];
  grandTotal: {
    totalTokens: number;
    estimatedCostUsd: number;
  };
}

// Public API. Assembles the Burn Summary envelope (schemaVersion 2).
// `period` selects the calendar window; window end and `generatedAt` are
// anchored to the same instant. Throws when `period` is unknown or no
// sessions fall in the window.
export async function buildEnvelope(
  inputs: CollectInputs,
  options: { generatedAt?: string } = {},
): Promise<Envelope> {
  const period: Period = inputs.period ?? "week";
  if (!PERIODS.includes(period)) {
    throw new Error("unknown period");
  }
  let now: Date;
  let generatedAt: string;
  if (options.generatedAt !== undefined) {
    const parsed = parseInstant(options.generatedAt);
    if (parsed === null) {
      throw new Error("invalid generatedAt");
    }
    now = parsed;
    generatedAt = formatGenerated(parsed);
  } else if (inputs.now) {
    now = inputs.now;
    generatedAt = formatGenerated(now);
  } else {
    now = new Date();
    generatedAt = formatGenerated(now);
  }
  const fallbackDay = generatedAt.slice(0, 10);
  const window = calendarWindow(period, now);
  const groups = await collectGroups({ ...inputs, now }, window);

  const sortedKeys = [...groups.__keys.entries()]
    .sort((a, b) => compareKey(a[1], b[1]));

  const rows: EnvelopeRow[] = [];
  let totalTokens = 0;
  let totalCost = 0;
  for (const [keyStr, keyTuple] of sortedKeys) {
    const grp = groups.get(keyStr)!;
    const [tool, model, phash] = keyTuple as [
      "claude" | "codex",
      string,
      string,
    ];
    const providerTable = tool === "claude"
      ? (MODEL_PRICING as PricingTable).claude
      : (MODEL_PRICING as PricingTable).codex;
    const { rate, confidence } = matchModel(
      providerTable as Record<string, ClaudeRate | CodexRate> & {
        _default: ClaudeRate | CodexRate;
      },
      model,
    );
    const priceRecord = rate as unknown as Record<string, number>;
    const breakdown = costBreakdown(grp.tokens, priceRecord);
    let rawCost = 0;
    for (const v of Object.values(breakdown)) rawCost += v;
    const cost = bankersRound(rawCost, 4);
    const tokenCount = schemaTokenCount(tool, grp.tokens as unknown as SessionTokens);
    const rowTotal = safeAdd(
      safeAdd(
        safeAdd(
          safeAdd(tokenCount.input, tokenCount.output),
          tokenCount.cacheRead,
        ),
        tokenCount.cacheWrite,
      ),
      tokenCount.cachedInput,
    );
    const days = [...grp.days].sort();
    rows.push({
      tool: TOOL_NAME[tool],
      model,
      tokenCount,
      estimatedCostUsd: cost,
      timestampBucket: days.length > 0 ? days[days.length - 1] : fallbackDay,
      sessionCount: grp.sessions,
      activeDays: days.length,
      projectHash: phash,
      verification: buildVerification(confidence),
    });
    totalTokens = safeAdd(totalTokens, rowTotal);
    const nextCost = totalCost + cost;
    totalCost = Number.isFinite(nextCost) ? nextCost : totalCost;
  }
  if (rows.length === 0) {
    throw new Error(`no sessions in period '${period}'`);
  }
  const periodWindow: PeriodWindow = window === null
    ? { period, since: null, until: null }
    : {
        period,
        since: formatGenerated(window[0]),
        until: formatGenerated(window[1]),
      };
  return {
    schemaVersion: "2",
    generatedAt,
    periodWindow,
    rows,
    grandTotal: {
      totalTokens,
      estimatedCostUsd: bankersRound(totalCost, 4),
    },
  };
}

// Internal helpers exported for unit-test parity checks. Not consumed by
// the UI layer.
export const __internal = {
  bankersRound,
  utcDay,
  parseInstant,
  isoWeekday,
  formatGenerated,
  calendarWindow,
  inWindow,
  schemaTokenCount,
  buildVerification,
  compareKey,
};
